lib/strings.nix: add meson-related utilities

AndersonTorres opened the pull request:
###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

AndersonTorres committed:
lib/strings.nix: add mesonOption utility function

And some friends, to help write Meson commandline invocations.

SuperSandro2000 commented on lib/strings.nix:
```diff
@@ -661,6 +661,40 @@ rec {
+       mesonFeature savage false
+       => "-Dsavage=disabled"
+  */
+  mesonFeature = feature: flag:
+    "-D${feature}=${if flag then "enabled" else "disabled"}";
```
Can we make this a wrapper of a third function which allows passing custom strings

```
mesonNotSureYet "feature" "auto"
```

AndersonTorres replied:
I like it, but `mesonPleaseINeedAGoodNameForThis`.

AndersonTorres replied:
Well, we have something like

```
-D${feature}=<true|false|enabled|disabled|customString>
```

Maybe `mesonOption feature value` to start with?

AndersonTorres committed:
muon: use mesonOption-related functions

This specific commit is intended to demonstrate a somewhat real world use case
for the new meson.* utilities defined in lib/strings.nix

SuperSandro2000 commented on pkgs/development/tools/build-managers/muon/default.nix:
```diff
@@ -81,14 +81,13 @@ stdenv.mkDerivation (finalAttrs: {
-      (muonConditionFlag "static" stdenv.targetPlatform.isStatic)
-      (muonFeatureFlag "docs" buildDocs)
-      (muonFeatureFlag "samurai" embedSamurai)
+      (muonBool "static" stdenv.targetPlatform.isStatic)
+      (muonEnable "docs" buildDocs)
+      (muonEnable "samurai" embedSamurai)
```
```suggestion
      (lib.mesonBool "static" stdenv.targetPlatform.isStatic)
      (lib.mesonEnable "docs" buildDocs)
      (lib.mesonEnable "samurai" embedSamurai)
```

SuperSandro2000 commented on pkgs/development/tools/build-managers/muon/default.nix:
```diff
@@ -81,14 +81,13 @@ stdenv.mkDerivation (finalAttrs: {
-      "-D${feature}=${if flag then "enabled" else "disabled"}";
-    muonConditionFlag = condition: flag:
-      "-D${condition}=${lib.boolToString flag}";
+    muonBool = lib.mesonBool;
+    muonEnable = lib.mesonEnable;
+
```
```suggestion
```

AndersonTorres replied:
I preferred this way because Muon commandline is not compatible with Meson.

SuperSandro2000 commented:
Please wait for someone that is more familiar with lib things than me but LGTM.

SuperSandro2000 commented on lib/strings.nix:
```diff
@@ -661,6 +661,61 @@ rec {
+  */
+  mesonEnable = feature: flag:
+    assert (isString feature);
+    assert (isBool flag);
```
```suggestion
    assert (builtins.isBool flag);
```

SuperSandro2000 commented on lib/strings.nix:
```diff
@@ -661,6 +661,61 @@ rec {
+  */
+  mesonBool = condition: flag:
+    assert (isString condition);
+    assert (isBool flag);
```
```suggestion
    assert (builtins.isBool flag);
```

AndersonTorres committed:
lib/default.nix: include mesonOption-related functions

To bring them to the scope when calling `lib`.

lucasew commented on lib/strings.nix:
```diff
@@ -661,6 +661,61 @@ rec {
+     @param flag The controlling flag of the condition
+
+     Example:
+       mesonBool hardened true
```
This `hardened` should be `"hardened"`

AndersonTorres replied:
Done

AndersonTorres committed:
slurp: 1.3.2 -> 1.4.0

AndersonTorres committed:
mpv: use meson* functions
